src/warp10-datasource.ts

```diff
@@ -260,24 +282,50 @@ export default class Warp10Datasource {
+    // current.text is the label. In case of multivalue, it is a string 'valueA + valueB'
+    // current.value is a string, depending on query output. In case of multivalue, it is an array of strings. array contains "$__all" if user selects All.
+    console.log("this.templateSrv.variables", this.templateSrv.variables)
+
     for (let myVar of this.templateSrv.variables) {
-      let value = myVar.current.text
-
-      if (myVar.current.value != null && (myVar.current.value === '$__all' || (myVar.current.value.length === 1 && myVar.current.value[0] === '$__all')))
-      {
-        if (myVar.allValue !== null)
-          value = myVar.allValue;
-        else
-          value = myVar.options.slice(1).map(e => e.text).join(" + ");
+      let value = myVar.current.value;
+
+      if (Array.isArray(value) && (value.length == 1 && value[0] === '$__all')) {
+        // User checked the "select all" checkbox
+        if (myVar.allValue && myVar.allValue !== "") {
+          // User also defined a custom value in the variable settings
+          let customValue: String = myVar.allValue;
+          wsHeader += `[ '${customValue}' ] '${myVar.name}' STORE\n`
+          // If user already starts its custom value with ~, remove it.
+          if (customValue.startsWith("~")) {
+            customValue = customValue.slice(1);
+          }
+          wsHeader += ` '~${customValue}' '${myVar.name}_wsregexp' STORE\n`
+        } else {
+          // if no custom all value is defined :
+          // it means we shall create a list of all the values in WarpScript from options, ignoring "$__all" special option value. 
+          let allValues: String[] = myVar.options.filter(o => o.value !== "$__all").map(o => o.value);
+          wsHeader += `[ ${allValues.map(s => `'${s}'`).join(" ")} ] '${myVar.name}' STORE\n`; // all is stored as string in generated WarpScript.
+          //also create a ready to use regexp, suffixed by _wsregexp
+          wsHeader += ` '~' $${myVar.name} REOPTALT + '${myVar.name}_wsregexp' STORE\n`
+        }
+      } else if (Array.isArray(value)) {
+        // user checks several choices
+        wsHeader += `[ ${value.map(s => `'${s}'`).join(" ")} ] '${myVar.name}' STORE\n`; // all is stored as string in generated WarpScript.
+        //also create a ready to use regexp, suffixed by _wsregexp
+        wsHeader += ` '~' $${myVar.name} REOPTALT + '${myVar.name}_wsregexp' STORE\n`
+      } else {
+        // no multiple selection, variable is the string. As type is lost by Grafana, there is no safe way to assume something different than a string here.
+        // User can use _long or _double suffixed variable
+        wsHeader += `'${value}' '${myVar.name}' STORE\n`;
+        //additionnal conversion for WarpScript illiterates users:
+        wsHeader += ` <% $${myVar.name} TOLONG %> <% NULL %> <% %> TRY '${myVar.name}_long' STORE\n`;
+        wsHeader += ` <% $${myVar.name} TODOUBLE %> <% NULL %> <% %> TRY '${myVar.name}_double' STORE\n`;
```

Comment:
Nice to have but I am not sure it is relevant. User can declare it if needed.

If we want it, we should apply it to multiple values also.

Reply:
Would prefer to drop it, for now, and come back to it later if needed.

Reply:
> Grafana did not handle numbers as values.
> I see in the code there was a special case for strings starting with 0. What was the use case ?
> That's why I added _long (and _double for consistency)

I think, that was to force escaping in JS templating (in case value was numeric).

Reply:
removed